Bee Chang Angel Martinez

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
#reading the data
df=pd.read_csv('GroceryStoreDataSet.csv')
df.head()

MILK,BREAD,BISCUIT
0  BREAD,MILK,BISCUIT,CORNFLAKES
1            BREAD,TEA,BOURNVITA
2           JAM,MAGGI,BREAD,MILK
3              MAGGI,TEA,BISCUIT
4            BREAD,TEA,BOURNVITA

In [ ]:
# missing a header we'll create one called 'Products' 
df=pd.read_csv('GroceryStoreDataSet.csv', names=['products'])
df.head() 


products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [ ]:
df.shape

(20, 1)

In [ ]:
#converting dataframe into a list of lists.
data=list(df["products"].apply(lambda x:x.split(",")))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [ ]:
#doing one hot encoding
#Each column represents an item and each row represent a record or a transaction. 
#items were given TRUE if it has occured and FALSE if it didnt 
#converted the str into a numeric value 

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True into numeric value
df=df.replace(False,0)
df=df.replace(True,1)

df
                

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  CORNFLAKES  JAM  MAGGI  MILK  \
0         1          0      1     0       0           0    0      0     1   
1         1          0      1     0       0           1    0      0     1   
2         0          1      1     0       0           0    0      0     0   
3         0          0      1     0       0           0    1      1     1   
4         1          0      0     0       0           0    0      1     0   
5         0          1      1     0       0           0    0      0     0   
6         0          0      0     0       0           1    0      1     0   
7         1          0      1     0       0           0    0      1     0   
8         0          0      1     0       0           0    1      1     0   
9         0          0      1     0       0           0    0      0     1   
10        1          0      0     1       1           1    0      0     0   
11        1          0      0     1       1           1    0      0     0   
12        0          1      0     0       1           0    0      0     0   
13        0          0      1     1       1           0    0      0     0   
14        1          0      1     0       0           0    0      0     0   
15        0          0      0     0       1           1    0      0     0   
16        0          1      1     0       0           0    0      0     0   
17        0          0      1     0       1           0    0      0     0   
18        0          0      1     0       1           0    0      0     0   
19        0          0      0     0       1           1    0      0     1   

    SUGER  TEA  
0       0    0  
1       0    0  
2       0    1  
3       0    0  
4       0    1  
5       0    1  
6       0    1  
7       0    1  
8       0    1  
9       0    0  
10      0    0  
11      0    0  
12      1    0  
13      0    0  
14      1    0  
15      1    0  
16      1    0  
17      1    0  
18      1    0  
19      0    1

In [ ]:
#we did .1 support value threshold 
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df=apriori(df,min_support=0.1,use_colnames=True)
df

support                             itemsets
0      0.35                            (BISCUIT)
1      0.20                          (BOURNVITA)
2      0.65                              (BREAD)
3      0.15                               (COCK)
4      0.40                             (COFFEE)
5      0.30                         (CORNFLAKES)
6      0.10                                (JAM)
7      0.25                              (MAGGI)
8      0.25                               (MILK)
9      0.30                              (SUGER)
10     0.35                                (TEA)
11     0.20                     (BISCUIT, BREAD)
12     0.10                      (BISCUIT, COCK)
13     0.10                    (COFFEE, BISCUIT)
14     0.15                (BISCUIT, CORNFLAKES)
15     0.10                     (BISCUIT, MAGGI)
16     0.10                      (MILK, BISCUIT)
17     0.10                       (TEA, BISCUIT)
18     0.15                   (BREAD, BOURNVITA)
19     0.10                   (SUGER, BOURNVITA)
20     0.10                     (TEA, BOURNVITA)
21     0.15                      (COFFEE, BREAD)
22     0.10                         (JAM, BREAD)
23     0.15                       (BREAD, MAGGI)
24     0.20                        (MILK, BREAD)
25     0.20                       (SUGER, BREAD)
26     0.20                         (TEA, BREAD)
27     0.15                       (COFFEE, COCK)
28     0.10                   (CORNFLAKES, COCK)
29     0.20                 (COFFEE, CORNFLAKES)
30     0.20                      (SUGER, COFFEE)
31     0.10                   (MILK, CORNFLAKES)
32     0.10                    (TEA, CORNFLAKES)
33     0.10                         (JAM, MAGGI)
34     0.20                         (TEA, MAGGI)
35     0.10               (MILK, BISCUIT, BREAD)
36     0.10              (COFFEE, BISCUIT, COCK)
37     0.10          (BISCUIT, CORNFLAKES, COCK)
38     0.10        (COFFEE, BISCUIT, CORNFLAKES)
39     0.10                (TEA, BISCUIT, MAGGI)
40     0.10              (TEA, BREAD, BOURNVITA)
41     0.10               (SUGER, COFFEE, BREAD)
42     0.10                  (JAM, BREAD, MAGGI)
43     0.10                  (TEA, BREAD, MAGGI)
44     0.10           (COFFEE, CORNFLAKES, COCK)
45     0.10  (COFFEE, BISCUIT, CORNFLAKES, COCK)

With the SUPPORT values above, you can try and match your own manual calculation. You notice how some are grouped with two. ALL items have been tested alone and in groups, but only these ones with the SUPPORT threshold 0.2 or higher values (i.e. other groups didn't make 0.2 therefore not shown).

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents                    consequents  antecedent support  \
0             (BISCUIT)                         (COCK)                0.35   
1                (COCK)                      (BISCUIT)                0.15   
2             (BISCUIT)                   (CORNFLAKES)                0.35   
3          (CORNFLAKES)                      (BISCUIT)                0.30   
4             (BISCUIT)                        (MAGGI)                0.35   
..                  ...                            ...                 ...   
99   (CORNFLAKES, COCK)              (COFFEE, BISCUIT)                0.10   
100            (COFFEE)    (BISCUIT, CORNFLAKES, COCK)                0.40   
101           (BISCUIT)     (COFFEE, CORNFLAKES, COCK)                0.35   
102        (CORNFLAKES)        (COFFEE, BISCUIT, COCK)                0.30   
103              (COCK)  (COFFEE, BISCUIT, CORNFLAKES)                0.15   

     consequent support  support  confidence       lift  leverage  conviction  
0                  0.15     0.10    0.285714   1.904762    0.0475       1.190  
1                  0.35     0.10    0.666667   1.904762    0.0475       1.950  
2                  0.30     0.15    0.428571   1.428571    0.0450       1.225  
3                  0.35     0.15    0.500000   1.428571    0.0450       1.300  
4                  0.25     0.10    0.285714   1.142857    0.0125       1.050  
..                  ...      ...         ...        ...       ...         ...  
99                 0.10     0.10    1.000000  10.000000    0.0900         inf  
100                0.10     0.10    0.250000   2.500000    0.0600       1.200  
101                0.10     0.10    0.285714   2.857143    0.0650       1.260  
102                0.10     0.10    0.333333   3.333333    0.0700       1.350  
103                0.10     0.10    0.666667   6.666667    0.0850       2.700  

[104 rows x 9 columns]

From above result, we say that the support is 0.2 calculated by dividing the number of transactions containing milk and bread by the total number of transactions.  The confidence level is 0.30 shows that out of all the transactions that contain bread 0.30 contain milk too.  The lift 1.23 tells us that milk is 1.23 times more likely to be bought by the customer who also buys bread.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents                    consequents  \
0                          (COCK)                      (BISCUIT)   
1                     (BOURNVITA)                        (BREAD)   
2                           (JAM)                        (BREAD)   
3                         (MAGGI)                        (BREAD)   
4                          (MILK)                        (BREAD)   
5                         (SUGER)                        (BREAD)   
6                          (COCK)                       (COFFEE)   
7                          (COCK)                   (CORNFLAKES)   
8                    (CORNFLAKES)                       (COFFEE)   
9                         (SUGER)                       (COFFEE)   
10                          (JAM)                        (MAGGI)   
11                        (MAGGI)                          (TEA)   
12                (MILK, BISCUIT)                        (BREAD)   
13              (COFFEE, BISCUIT)                         (COCK)   
14                 (COFFEE, COCK)                      (BISCUIT)   
15                (BISCUIT, COCK)                       (COFFEE)   
16                         (COCK)              (COFFEE, BISCUIT)   
17          (BISCUIT, CORNFLAKES)                         (COCK)   
18                (BISCUIT, COCK)                   (CORNFLAKES)   
19             (CORNFLAKES, COCK)                      (BISCUIT)   
20                         (COCK)          (BISCUIT, CORNFLAKES)   
21              (COFFEE, BISCUIT)                   (CORNFLAKES)   
22          (BISCUIT, CORNFLAKES)                       (COFFEE)   
23                 (TEA, BISCUIT)                        (MAGGI)   
24               (BISCUIT, MAGGI)                          (TEA)   
25               (TEA, BOURNVITA)                        (BREAD)   
26             (BOURNVITA, BREAD)                          (TEA)   
27                (COFFEE, BREAD)                        (SUGER)   
28                   (JAM, BREAD)                        (MAGGI)   
29                   (JAM, MAGGI)                        (BREAD)   
30                 (BREAD, MAGGI)                          (JAM)   
31                          (JAM)                 (BREAD, MAGGI)   
32                 (BREAD, MAGGI)                          (TEA)   
33                 (COFFEE, COCK)                   (CORNFLAKES)   
34             (CORNFLAKES, COCK)                       (COFFEE)   
35                         (COCK)           (COFFEE, CORNFLAKES)   
36  (COFFEE, BISCUIT, CORNFLAKES)                         (COCK)   
37        (COFFEE, BISCUIT, COCK)                   (CORNFLAKES)   
38     (COFFEE, CORNFLAKES, COCK)                      (BISCUIT)   
39    (BISCUIT, CORNFLAKES, COCK)                       (COFFEE)   
40              (COFFEE, BISCUIT)             (CORNFLAKES, COCK)   
41                 (COFFEE, COCK)          (BISCUIT, CORNFLAKES)   
42          (BISCUIT, CORNFLAKES)                 (COFFEE, COCK)   
43                (BISCUIT, COCK)           (COFFEE, CORNFLAKES)   
44             (CORNFLAKES, COCK)              (COFFEE, BISCUIT)   
45                         (COCK)  (COFFEE, BISCUIT, CORNFLAKES)   

    antecedent support  consequent support  support  confidence       lift  \
0                 0.15                0.35     0.10    0.666667   1.904762   
1                 0.20                0.65     0.15    0.750000   1.153846   
2                 0.10                0.65     0.10    1.000000   1.538462   
3                 0.25                0.65     0.15    0.600000   0.923077   
4                 0.25                0.65     0.20    0.800000   1.230769   
5                 0.30                0.65     0.20    0.666667   1.025641   
6                 0.15                0.40     0.15    1.000000   2.500000   
7                 0.15                0.30     0.10    0.666667   2.222222   
8                 0.30                0.40     0.20    0.666667   1.666667   
9                 0.30                0.40   

From above result, let's take a look at the first row items: milk and bread. 
The two items have the support of 0.20, where milk is 0.25 and bread is 0.65. 
confidence is 0.8, this implies that if a person bought milk then there is 80% bread is also bought. There are also positive correlations as we can see from the positive values of lift, leverage, and conviction.